Questionnaire: Drag and Drop "Sorting" question

lamtranb opened the pull request:
Hi, 

We are from the Open University and would like to do this change:

Create a new question type Drag and Drop "Sorting".

Could you please review this commit and inform for any issue?

Thanks.

lamtranb commented:
Hi @mchurchward 

I have rebased the code. Could you review this PR?

Thanks.

mchurchward commented:
Hi. Thanks for this... some review comments:

1. Is the new response table necessary? It looks like it could use the standard "questionnaire_response_text" table.
2. What is the purpose of the new "mobile_questionnaire_init" function in the mobile.php class?
3. Can you explain the changes made to "db/mobile.php"?
4. It appears you have replaced the "styles_app.css" with the new "mobile_css.php" file. Given that, should the "styles_app.css" be deleted?
5. Can the "questionnaire_prep_answer_sort_question" function in locallib be refactored into one of the classes? Maybe the specific response class?
6. Can you add the new question type to the unit test for csv exporting?

mchurchward reviewed (commented):
Hi. Thanks for this... some review comments:

1. Is the new response table necessary? It looks like it could use the standard "questionnaire_response_text" table.
2. What is the purpose of the new "mobile_questionnaire_init" function in the mobile.php class?
3. Can you explain the changes made to "db/mobile.php"?
4. It appears you have replaced the "styles_app.css" with the new "mobile_css.php" file. Given that, should the "styles_app.css" be deleted?
5. Can the "questionnaire_prep_answer_sort_question" function in locallib be refactored into one of the classes? Maybe the specific response class?
6. Can you add the new question type to the unit test for csv exporting?

Commit:
Questionnaire: Drag and Drop "Sorting" question #323936

lamtranb commented:
Hi @mchurchward, 

I updated the code.

> 1. Is the new response table necessary? It looks like it could use the standard "questionnaire_response_text" table.

Yes, because we store a string example "1,2,3" which are sorting choices. And later we can transform it into array, so it has different behavior from questionnaire_response_text.

> 2. What is the purpose of the new "mobile_questionnaire_init" function in the mobile.php class?

This is used for mobile app. It called the JS file: appjs/latest/mobile_view_activity.js. And sadly, we did not support for ionic 3 yet.

> 3. Can you explain the changes made to "db/mobile.php"?

The changes we made in this file is we called the PHP method mobile_questionnaire_init (point 2).

> 4. It appears you have replaced the "styles_app.css" with the new "mobile_css.php" file. Given that, should the "styles_app.css" be deleted?

Thanks for correct me. I added some code into styles_app.css and removed mobile.css.

> 5. Can the "questionnaire_prep_answer_sort_question" function in locallib be refactored into one of the classes? Maybe the specific response class?

Thanks for your suggestion. I moved all sort methods from locallib to responsetype\sorting.

> 6. Can you add the new question type to the unit test for csv exporting?

Of course. I added some UTs.

Could you please review the code again?

Many thanks

rezeau commented:
Cannot test and review because this PR must be re-based against current MOODLE_400_STABLE first.

mchurchward committed:
Setting up release for Moodle 4.1 onward.

lucaboesch committed:
Allow localized answer options to be displayed correctly in conditions. (#449)

* Resized the monologo.svg and png to 24x24 px (#477)

* Fixing issues from plugins database checks (#483)

* Removing illegal console commands. Script may not be needed.

* Fixing illegal line length.

* Fixing docblocks and unused variables.

* Fixing css spaces.

* Adding a true return.

* Fixing JSON context.

* Allow localized answer options to be displayed correctly in conditions.

---------

Co-authored-by: FelixDiLenarda <92793568+FelixDiLenarda@users.noreply.github.com>
Co-authored-by: Mike Churchward <mike.churchward@poetopensource.org>

toanlamt committed:
Questionnaire: improved accessibility for essay box type (#506)

* The text editor should be programmatically associated with the question

Co-authored-by: toanlam <toanlamt1@nashtechglobal.com>

mchurchward committed:
Removing 4.3 for now.

tailetan committed:
Questionnaire\Accessibility\Slider: The slider values should be associated with the labels and the slider should be programmatically associated with the question (#495)

* Questionnaire: More meaningful error than nopermissions (#488)

Co-authored-by: Emanoil Manoylov <em22699@open.ac.uk>

* Questionnaire\Accessibility\Slider: The slider values should be associated with the labels #674064

---------

Co-authored-by: emanoylov <80328987+emanoylov@users.noreply.github.com>
Co-authored-by: Emanoil Manoylov <em22699@open.ac.uk>
Co-authored-by: tai.letan <tlt84@open.ac.uk>

lamtranb committed:
Questionnaire\Accessibility\Rate: table does not have a programmatically associated caption (#497)

* Questionnaire: More meaningful error than nopermissions (#488)

Co-authored-by: Emanoil Manoylov <em22699@open.ac.uk>

* Questionnaire\Accessibility\Rate: table does not have a programmatically associated caption

---------

Co-authored-by: emanoylov <80328987+emanoylov@users.noreply.github.com>
Co-authored-by: Emanoil Manoylov <em22699@open.ac.uk>
Co-authored-by: lamtranb <ltb58@open.ac.uk>

tailetan committed:
Questionnaire\Accessibility\Numeric: Instructions not programmatically associated with field #672028 (#496)

Co-authored-by: tai.letan <tlt84@open.ac.uk>

tailetan commented:
I have already rebased this with MOODLE_401_STABLE.
FYI: @rezeau, @mchurchward

lamtranb committed:
Questionnaire\Accessibility\Rate: The form controls within the table are not accessible (#505)

Co-authored-by: lamtranb <ltb58@open.ac.uk>

emanoylov committed:
Questionnaire/Accessibility: Check boxes missing group label (#501)

Co-authored-by: Emanoil Manoylov <em22699@open.ac.uk>

opitz committed:
Fixing issues #502 and #507 (#508)

* now using core_external\external_api - fixes #502

* using valid $required parameter VALUE_REQUIRED - fixes #507

---------

Co-authored-by: Matthias Opitz <m.opitz@ucl.ac.uk>

tailetan commented:
I have already rebased this with MOODLE_401_STABLE.
FYI: @mchurchward

mchurchward committed:
Adding Moodle 4.3beta to the mix.

lamtranb commented:
Dear @tailetan , this branch has conflicts again. Could you please resolve it?

tailetan commented:
@lamtranb, I've just rebased the latest code. Thank you
